src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/importer.h"
+
+#include <algorithm>
+#include <cmath>
+#include <fstream>
+#include <libxml/uri.h>
+#include <sstream>
+#include <stdexcept>
+
+#include "libcellml/component.h"
+#include "libcellml/importsource.h"
+#include "libcellml/model.h"
+#include "libcellml/parser.h"
+#include "libcellml/reset.h"
+#include "libcellml/units.h"
+#include "libcellml/variable.h"
+
+#include "namespaces.h"
+#include "utilities.h"
+#include "xmldoc.h"
+#include "xmlutils.h"
+
+namespace libcellml {
+
+/**
+ * @brief The Importer::ImporterImpl struct.
+ *
+ * This struct is the private implementation struct for the Importer class.  Separating
+ * the implementation from the definition allows for greater flexibility when
+ * distributing the code.
+ */
+struct Importer::ImporterImpl
+{
+    Importer *mImporter = nullptr;
+
+    ImportLibrary mLibrary;
+    std::vector<std::string> mLibraryIndex;
+    std::vector<std::pair<std::string, std::string>> mExternals;
+
+    bool resolveImport(const ImportedEntityPtr &importedEntity,
+                       const std::string &destination,
+                       const std::string &typeString,
+                       const std::string &baseFile,
+                       std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool resolveComponentImports(const ComponentEntityPtr &parentComponentEntity,
+                                 const std::string &baseFile,
+                                 std::vector<std::tuple<std::string, std::string, std::string>> &history);
+
+    bool doResolveImports(ModelPtr &model, const std::string &baseFile,
+                          std::vector<std::tuple<std::string, std::string, std::string>> &history);
+};
+
+Importer::Importer()
+    : mPimpl(new ImporterImpl())
+{
+    mPimpl->mImporter = this;
+}
+
+ImporterPtr Importer::create() noexcept
+{
+    return std::shared_ptr<Importer> {new Importer {}};
+}
+
+Importer::~Importer()
+{
+    delete mPimpl;
+}
+
+/**
+ * @brief Resolve the path of the given filename using the given base.
+ *
+ * Resolves the full path to the given @p filename using the @p base.
+ *
+ * This function is only intended to work with local files.  It may not
+ * work with bases that use the 'file://' prefix.
+ *
+ * @param filename The @c std::string relative path from the base path.
+ * @param base The @c std::string location on local disk for determining the full path from.
+ *
+ * @return The full path from the @p base location to the @p filename
+ */
+std::string resolvePath(const std::string &filename, const std::string &base)
+{
+    // We can be naive here as we know what we are dealing with.
+    std::string path = base.substr(0, base.find_last_of('/') + 1) + filename;
+    return path;
+}
```

Comment:
This method is only used once ([here](https://github.com/cellml/libcellml/pull/608/files#diff-460f06cc38b9cccaf120da540eda32a1R185)), so would move that code there.

Reply:
7b7ad654